include/ccf/ds/pal.h

```diff
@@ -81,6 +94,106 @@ namespace ccf
       info.peak_allocated_heap_size = 0;
       return true;
     }
+
+    static QuoteInfo generate_quote(attestation_report_data&& report_data)
+    {
+      QuoteInfo node_quote_info = {};
+      auto is_sev_snp = access(SEV_SNP_DEVICE, F_OK) == 0;
+
+      // If there is no SEV-SNP device, assume we are using insecure virtual quotes
+      if (!is_sev_snp) {
+        node_quote_info.format = QuoteFormat::insecure_virtual;
+      }
+
+      else {
+        node_quote_info.format = QuoteFormat::amd_sev_snp_v1;
+        int fd = open(SEV_SNP_DEVICE, O_RDWR | O_CLOEXEC);
+        if (fd < 0)
+        {
+          throw std::logic_error("Failed to open \"/dev/sev\"");
+        }
+
+        msg_report_req req = {};
+        msg_report_rsp resp = {};
+
+        // Arbitrary report data
+        memcpy(req.report_data, report_data.data(), attestation_report_data_size);
+
+        sev_snp_guest_request payload = {
+            .req_msg_type = SNP_MSG_REPORT_REQ,
+            .rsp_msg_type = SNP_MSG_REPORT_RSP,
+            .msg_version = 1,
+            .request_len = sizeof(req),
+            .request_uaddr = (uint64_t)(void*)&req,
+            .response_len = sizeof(resp),
+            .response_uaddr = (uint64_t)(void*)&resp,
+            .error = 0
+        };
+
+        int rc = ioctl(fd, SEV_SNP_GUEST_MSG_REPORT, &payload);
+        if (rc < 0)
+        {
+          CCF_APP_FAIL("IOCTL call failed: {}", strerror(errno));
+          CCF_APP_FAIL("Payload error: {}", payload.error);
+          throw std::logic_error(
+            "Failed to issue ioctl SEV_SNP_GUEST_MSG_REPORT");
+        }
+
+        auto quote = reinterpret_cast<uint8_t*>(&resp.report);
+        node_quote_info.quote.assign(quote, quote + resp.report_size);
+        // TODO: Get endorsements
+      }
+      return node_quote_info;
+    }
+
+    static void verify_quote(
+      const QuoteInfo& quote_info,
+      attestation_measurement& unique_id,
+      attestation_report_data& report_data)
+    {
+      auto is_sev_snp = access(SEV_SNP_DEVICE, F_OK) == 0;
+
+      if (quote_info.format == QuoteFormat::insecure_virtual)
+      {
+        if (is_sev_snp) {
+          throw std::logic_error(
+            "Cannot verify virtual quote if node is SEV-SNP");
+        }
+        unique_id = {};
+        report_data = {};
+      }
+
+      else if (quote_info.format == QuoteFormat::amd_sev_snp_v1) {
+        if (!is_sev_snp) {
+          throw std::logic_error(
+            "Cannot verify SEV-SNP quote if node is virtual");
+        }
+        attestation_report quote = *reinterpret_cast<const attestation_report*>(quote_info.quote.data());
+        std::copy(
+          std::begin(quote.report_data),
+          std::begin(quote.report_data) + attestation_report_data_size,
```

Comment:
This needs fixing. We shouldn't truncate the measurement on AMD but use the full 64 bytes. I guess `attestation_measurement` has to become a `std::vector<uint8_t>` and we should check its size based on the type of platform type later.

Reply:
This is the report data rather than the measurement, the report data if I understand correctly is provided by us and is a 32 byte hash of the nodes public key? In which case I don't think there is any issues with truncating down to 32 bytes right?

The measurement uses all 48 bytes for AMD as of this commit: https://github.com/microsoft/CCF/pull/4106/commits/48b6eabca9ca5801916b4aab826de5913345d573